### DISCUSSION TITLE
Invokable actions

### DISCUSSION THREAD

**Discussion by demobiel** [filamentphp/filament#6046] [Help]
Hi

I would like to use invokable classes to use for my actions. Something like this:
````
public static function table(Table $table): Table
    {
        return $table
            ->actions([
                Tables\Actions\EditAction::make(),
                Tables\Actions\Action::make('test')->action(\App\Actions\Test\MyAction::class)
            ])
````
However this is not working throwing the following error.
```
Failed to create closure from callable: function "App\Actions\Test\MyAction" not found or invalid function name
```
What I don't understand is that in file Filament\Support\Actions\Concerns\HasAction on line 23 there is actually a check if a string passed to an action is a class (looks to me that it should be an invokable class).

````
        if (is_string($action)) {
            $action = Closure::fromCallable(class_exists($action) ? $action : [$this->getLivewire(), $action]);
        }
````
Changing this code to new $action() actually makes it work.
````
        
if (is_string($action)) {
    $action = Closure::fromCallable(class_exists($action) ? new $action() : [$this->getLivewire(), $action]);
}
`````
Am I misunderstanding the use of this? Or is this a bug?

Kind regards

Andries

**Comment by demobiel** [comment]
After some more testing it also works if I do like this:

````
                Tables\Actions\Action::make('test')
                    ->action(function (Account $record ) {
                        $action = new MyAction($record->id);
                        return $action();
                    })
````
but this actually looks like a lot of "extra code".

**Reply by danharrin** [reply]
You can use DI:

```php
->action(fn (Account $record, MyAction $myAction) => $myAction($record->id)())
```

**Reply by demobiel** [reply]
Very good! I didn't think about that! Thanks for your help!

However I still wonder in which scenario you would want to pass a string that represents a class to the action method.

**Reply by danharrin** [reply]
The string actually represents a Livewire action method, not an invokable class in this case.

**Comment by demobiel** [comment]
OK! 

I have been playing around with it last night and I wonder if it would be interesting to have an invokable class that you could pass? It should implement a "handle" method that receives the resources as parameter.

Some kind of abstract class looking like this:
````

abstract InvokableAction
{
    public function __invoke(): void
    {
        $this->handle(func_get_args()[4]);
    }

    public function handle($resource){
        //
    }
}

````
It comes close to using a Closure as action that passes the resource as parameter. But it allows for more abstraction of your code. I have it now working, but need to rework some stuff, but if this would be interesting for the framework I am willing to do a PR for this.

**Reply by danharrin** [reply]
I don't think the handle() method is necessary, I'd just use __invoke(). But yeah, a PR to check class_exists() on an action string would be good.

**Reply by demobiel** [reply]
That test already exists in the Filament\Support\Actions\Concerns\HasAction concern on line 23, however that returns just the string and not the instantiated class.
````
$action = Closure::fromCallable(class_exists($action) ? $action : [$this->getLivewire(), $action]);
````
Problem is that I don't find any usage why it should use a string instead of an instantiated object, as Closure::fromCallable() only excepts callable parameter. That is why I asked in the beginning if this was a bug or not. Simply changing this line to return the instantiated object actually solves it.
````
$action = Closure::fromCallable(class_exists($action) ? new $action() : [$this->getLivewire(), $action]);
````
If this is what you want, and you think this is a good idea, I want to make a PR for this. However I did not test how this would behave if used somewhere else, until now I only used it for actions on my table.

**Reply by danharrin** [reply]
it should be `app($action)` instead of `new $action()` but yeah, a PR to fix would be good

**Reply by demobiel** [reply]
good point!
